orchagent/vnetorch.cpp

```diff
@@ -1967,6 +2228,227 @@ void VNetRouteOrch::updateVnetTunnel(const BfdUpdate& update)
     }
 }
 
+void VNetRouteOrch::updateVnetTunnelCustomMonitor(const MonitorUpdate& update)
+{
+    SWSS_LOG_ENTER();
+// This function recieves updates from the MonitorOrch for the endpoints state.
+// Based on the state of the endpoints for a pirticuar route, this function attempts
+// to construct the primary next hop group. if it fails to do so,it attempts to create
+// the secondary next hop group. After that it applies the next hop group and deletes
+// the old next hop group.
+// This function is also called in the case when the route configuration is updated to
+// apply the new next hop group. In this case, the caller sets the state to
+// MONITOR_SESSION_STATE_UNKNOWN and config_update and updateRoute are set to true.
+// This function should never recieve MONITOR_SESSION_STATE_UNKNOWN from MonitorOrch.
+
+    auto prefix = update.prefix;
+    auto state = update.state;
+    auto monitor = update.monitor;
+    auto vnet = update.vnet;
+    bool updateRoute = false;
+    bool config_update = false;
+    if (state != MONITOR_SESSION_STATE_UNKNOWN)
+    {
+        monitor_info_[vnet][prefix][monitor].state = state;
+    }
+    else
+    {
+        // we are coming here as a result of route config update. We need to repost the route if applicable.
+        updateRoute = true;
+        config_update = true;
+    }
+
+    auto route = syncd_tunnel_routes_[vnet].find(prefix);
+    if (route == syncd_tunnel_routes_[vnet].end())
+    {
+        SWSS_LOG_ERROR("Unexpected! Monitor Update for absent route.");
+        return;
+
+    }
+    auto *vrf_obj = vnet_orch_->getTypePtr<VNetVrfObject>(vnet);
+    set<sai_object_id_t> vr_set;
+
+    auto l_fn = [&] (const string& vnet) {
+        auto *vnet_obj = vnet_orch_->getTypePtr<VNetVrfObject>(vnet);
+        sai_object_id_t vr_id = vnet_obj->getVRidIngress();
+        vr_set.insert(vr_id);
+    };
+
+    l_fn(vnet);
+
+    auto primary = route->second.primary;
+    auto secondary = route->second.secondary;
+    auto active_nhg = route->second.nhg_key;
+    NextHopGroupKey nhg_custom("", true);
+    sai_ip_prefix_t pfx;
+    copy(pfx, prefix);
+    NextHopGroupKey nhg_custom_primary = getActiveNHSet( vnet, primary, prefix);
+    NextHopGroupKey nhg_custom_secondary = getActiveNHSet( vnet, secondary, prefix);
```

Comment:
This can be moved further down to avoid an unnecessary call in case primary nhg is not empty.